### ISSUE TITLE
Unable to Replicate Llama Compression Rate 

### ISSUE THREAD

**Issue by mxtsai** [google-deepmind/language_modeling_is_compression#19]
Hi,

Thank you for your interesting work. I have been trying to replicate your implementation using the most recent Llama models, but I am unable to achieve a compression rate as you have shown in your paper (<10%). I have seen some inquires regarding the Llama implementation. Following what you have described in the paper (Appendix B.2) and your responses regarding the implementation in some on other questions, I adjusted the `predict_fn` to be something similar to the following:

```
model_path = "meta-llama/Llama-3.1-8B"
tokenizer = AutoTokenizer.from_pretrained(model_path)
model = AutoModelForCausalLM.from_pretrained(model_path, device_map="cuda", torch_dtype=torch.bfloat16)

def predict_fn(sequence_array):
  with torch.no_grad():
    sequence_str = ''.join([chr(x) for x in sequence_array])  # convert from bytes back to string
    inputs = tokenizer(sequence_str, return_tensors="pt").input_ids.cuda()
    outputs = model(inputs, return_dict=True, output_hidden_states=True)
    logits = outputs.logits
    log_probs = torch.nn.functional.log_softmax(logits, dim=-1).float()
  return log_probs.cpu().numpy()
```

I kept the Arithmetic Coder intact, and used `meta-llama/Llama-3.1-8B` as the model of choice. However, I was only able to achieve a compression rate of ~70%.

Would it be possible if you can clarify whether there is an issue with my implementation? Thank you in advance.

**Comment by anianruoss** [comment]
Thank you for your interest in our work!

A few questions:

- On which dataset do you achieve a compression rate of ~70%?
- What sequence length are did you use?
- What log loss do you achieve on the same dataset?

**Comment by mxtsai** [comment]
Thank you for the prompt response.

I tested compression on the `enwik9` dataset. Prior to using Llama2, I used your `train.py` and was able to train a Transformer model to a log loss of around ~2800 (I don't remember the exact numbers as this was a few days ago). That Transformer model was able to achieve a compression rate of 17.7. The sequence length was 2048.

After achieving this using the trained Transformer model, I then attempted to swap out the Transformer model with Llama, and I was expecting a lower compression rate. Unfortunately, I wasn't able to replicate that with the above code (I tried following your suggestions in the other issues, and tried to follow the descriptions in the Appendix). 

Not sure if any part of my implementation is incorrect?

**Comment by anianruoss** [comment]
What is the log loss of your Llama implementation on this dataset?

You can sanity-check your implementation by computing the lower bound on the compression rate as follows
```
compressed_length = -math.ceil(-log_loss / math.log(2) / 8) + 1
compression_rate = compressed_length / sequence_length
```

For example, for your transformer with a log loss of around 2800, we obtain `compressed_length = 505`, and, therefore, a compression rate of 24.7.

**Comment by ACMCMC** [comment]
Hello,

I'm unable to replicate this as well. Would it be possible to ask for the script you used to get the Llama-2 results presented in the paper? That would be helpful :)

Edit: just seen that this was requested in #14. The answer there was that "Unfortunately, we will not be releasing our implementation of the Llama 2 compressor". But I find it hard to understand why this is the case - especially when it's such a simple implementation. Reproducibility is key in science, and I would really appreciate being able to replicate your work 🥇 Maybe there's a chance you could reconsider?

Thanks!

**Comment by anianruoss** [comment]
Unfortunately, this is outside of our control, sorry. I'm happy to help you debug your implementation. Did you try the above sanity check?
